core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -590,6 +627,48 @@ class TiRegionPartitioner(regions: List[TiRegion]) extends Partitioner {
   }
 }
 
+class keyWithDupInfo(var key: SerializableKey, var oldRow: TiRow) extends Serializable {
+  var oldHandle: Long = -1L
+  def setOldHandle(handle: Long): Unit = this.oldHandle = handle
+
+  override def equals(that: Any): Boolean =
+    that match {
+      case another: keyWithDupInfo =>
+        another.key.equals(this.key)
+      case _ => false
+    }
+
+  override def hashCode(): Int = util.Arrays.hashCode(key.bytes)
+}
+
+class toBeCheckedRow(val row: TiRow,
+                     val handleKey: keyWithDupInfo,
+                     val indexKeys: List[keyWithDupInfo])
+    extends Serializable {
+
+  //  val idxId: Long
+  // if handle is conflict or any key in indices key conflict with each other
+  // then two row equals
+  override def equals(that: Any): Boolean =
+    that match {
+      case row: toBeCheckedRow =>
+        if (row == this) return true
+        if (indexKeys.isEmpty && row.indexKeys.isEmpty) return false
+        val mayConflict = row.indexKeys.toSet
+        this.indexKeys.forall(mayConflict)
+      case _ =>
+        false
+    }
+
+  override def hashCode(): Int = {
+    var hash = 7
+    indexKeys.foreach { kv =>
+      hash += util.Arrays.hashCode(kv.key.bytes)
+    }
+    hash
```

Comment:
what about use `Objects.hash(indexKeys)`

Reply:
It is basically the same thing, but I can replace it with Objects.